L1Trigger/L1THGCal/src/backend/HGCalMulticlusteringImpl.cc

```diff
@@ -170,6 +192,314 @@ void HGCalMulticlusteringImpl::clusterizeDBSCAN( const std::vector<edm::Ptr<l1t:
+                    float y_seed = ROverZ_seed*sin(phi_seed);
+                    seedPositions.emplace_back(x_seed,y_seed,z_side);
+                }
+
+            }
+
+        }
+
+    }
+
+    return seedPositions;
+
+}
+
+
+
+std::vector<l1t::HGCalMulticluster> HGCalMulticlusteringImpl::clusterSeedMulticluster(const std::vector<edm::Ptr<l1t::HGCalCluster>> & clustersPtrs,
+										      const std::vector<GlobalPoint> seeds){
+
+
+    std::map<int,l1t::HGCalMulticluster> mapSeedMulticluster;
+    std::vector<l1t::HGCalMulticluster> multiclustersTmp;
+
+    for(std::vector<edm::Ptr<l1t::HGCalCluster>>::const_iterator clu = clustersPtrs.begin(); clu != clustersPtrs.end(); ++clu){
+
+        HGCalDetId cluDetId( (**clu).detId() );
+        int z_side = cluDetId.zside();
+
+        double minDist = dr_;
+        int targetSeed = -1;
+
+        for( unsigned int iseed=0; iseed<seeds.size(); iseed++ ){
+
+            if( z_side*seeds[iseed].z()<0) continue;
+
+            double d = this->dR(**clu, seeds[iseed]);
+
+            if(d<minDist){
+                minDist = d;
+                targetSeed = iseed;
+            }
+
+        }
+
+        if(targetSeed<0) continue;
+
+        if(mapSeedMulticluster[targetSeed].size()==0){
+            l1t::HGCalMulticluster newMclu(*clu);
+            mapSeedMulticluster[targetSeed] = newMclu;
+        }
+        else mapSeedMulticluster[targetSeed].addConstituent(*clu);
+
+    }
+
+    for(auto mclu : mapSeedMulticluster) multiclustersTmp.emplace_back(mclu.second);
+
+    return multiclustersTmp;
+
+}
+
+
+
+
+void HGCalMulticlusteringImpl::clusterizeHistoMax( const std::vector<edm::Ptr<l1t::HGCalCluster>> & clustersPtrs,
+						   l1t::HGCalMulticlusterBxCollection & multiclusters,
+						   const HGCalTriggerGeometryBase & triggerGeometry)
+{
+
+    /* put clusters into an r/z x phi histogram */
+    std::map<std::vector<int>,float> histoCluster = fillHistoClusters(clustersPtrs); //key[0] = z.side(), key[1] = bin_R, key[2] = bin_phi, content = MIPTs summed along depth
+
+    /* smoothen along the phi direction + normalize each bin to same area */
+    std::map<std::vector<int>,float> smoothPhiHistoCluster = fillSmoothPhiHistoClusters(histoCluster,binsSumsHisto_);
+
+    /* smoothen along the r/z direction */
+    std::map<std::vector<int>,float> smoothRPhiHistoCluster = fillSmoothRPhiHistoClusters(histoCluster);
+
+    /* seeds determined with local maximum criteria */
+    std::vector<GlobalPoint> seedPositions = computeMaxSeeds(smoothRPhiHistoCluster);
+
+    /* clusterize clusters around seeds */
+    std::vector<l1t::HGCalMulticluster> multiclustersTmp = clusterSeedMulticluster(clustersPtrs,seedPositions);
+
+    /* making the collection of multiclusters */
+    finalizeClusters(multiclustersTmp, multiclusters, triggerGeometry);
+
+}
+
+
+
+
+
+void HGCalMulticlusteringImpl::clusterizeHistoThreshold( const std::vector<edm::Ptr<l1t::HGCalCluster>> & clustersPtrs,
```

Comment:
This function is exactly the same as the previous one except one line. They could be merged to avoid code duplication.